File: transform_page.py
import sys

with open('app/dashboard/competitions/page.tsx', 'r', encoding='utf-8') as f:
    lines = f.readlines()

# find return start
return_idx = -1
for i, line in enumerate(lines):
    if line.startswith('    return (') and lines[i+1].strip().startswith('<Container'):
        return_idx = i
        break

# find detail panel start
panel_start = -1
for i, line in enumerate(lines):
    if line.strip() == '{selectedCompetition && (' and lines[i+1].strip().startswith('<Paper shadow="sm"'):
        panel_start = i
        break

# find detail panel end
panel_end = -1
for i in range(panel_start, len(lines)):
    if line.strip() == '{/* Create / Edit Modal */}':
        pass
    if '            {/* Create / Edit Modal */}' in lines[i]:
        panel_end = i
        # backtrack to `            )}` 
        while not lines[panel_end-1].strip() == ')}':
            panel_end -= 1
        panel_end -= 1 # The line with `            )}`
        break

if return_idx == -1 or panel_start == -1 or panel_end == -1:
    print(return_idx, panel_start, panel_end)
    sys.exit(1)

# we will create a function const DetailPanelComponent = () => { ... }
# taking the block from panel_start+1 to panel_end-1
# Wait, panel_start is `{selectedCompetition && (`. The actual JSX is panel_start+1 to panel_end-1
jsx_lines = lines[panel_start+1:panel_end]

# Change the wrapping Paper
for i, l in enumerate(jsx_lines):
    if l.strip().startswith('<Paper shadow="sm" radius="md"'):
        jsx_lines[i] = '            <Paper shadow={viewMode === \\'calendar\\' ? "sm" : "none"} radius="md" p={viewMode === \\'calendar\\' ? "sm" : 0} mt={viewMode === \\'calendar\\' ? "sm" : 0} className={viewMode === \\'calendar\\' ? styles.detailPanel : \\'\\'} pos="relative">\\n'
        break

detail_component_lines = [
    '    const DetailPanelComponent = () => {\\n',
    '        if (!selectedCompetition) return null;\\n',
    '        return (\\n',
    '            <Box>\\n' # Wrap in fragment just in case
] + jsx_lines + [
    '            </Box>\\n',
    '        );\\n',
    '    };\\n\\n'
]

# Delete competition detail header
header_start = -1
header_end = -1
for i in range(return_idx, panel_start):
    if lines[i].strip() == '{/* Competition Detail Section */}':
        header_start = i
        for j in range(i, panel_start):
            if lines[j].strip() == ')}':
                header_end = j + 1
                break
        break

# Reassemble
new_lines = lines[:return_idx] + detail_component_lines + lines[return_idx:header_start] + [
    '            {viewMode === \\'calendar\\' && <DetailPanelComponent />}\\n'
] + lines[panel_end+1:]

with open('app/dashboard/competitions/page.tsx', 'w', encoding='utf-8') as f:
    f.writelines(new_lines)
print("SUCCESS")

File: transform_page.js
const fs = require('fs');

const path = 'app/dashboard/competitions/page.tsx';
let content = fs.readFileSync(path, 'utf8');

// Normalize newlines to \n
content = content.replace(/\\r\\n/g, '\\n');

const panelIdx = content.indexOf('className={styles.detailPanel}');
const detailPanelStart = content.lastIndexOf('{selectedCompetition && (', panelIdx);

if (detailPanelStart === -1) {
    console.error("Could not find detailPanelStart");
    process.exit(1);
}

const detailPanelEndStr = '            {/* Create / Edit Modal */}';
const detailPanelEnd = content.indexOf(detailPanelEndStr, detailPanelStart);

if (detailPanelEnd === -1) {
    console.error("Could not find detailPanelEnd");
    process.exit(1);
}

const blockEnd = content.lastIndexOf('\\n', detailPanelEnd);

const detailPanelBlockStr = content.substring(detailPanelStart, blockEnd);

let modifiedDetailPanelBlock = detailPanelBlockStr
    .replace('<Paper shadow="sm" radius="md" p="sm" mt="sm" className={styles.detailPanel} pos="relative">',
        `<Paper shadow={viewMode === 'calendar' ? "sm" : "none"} radius="md" p={viewMode === 'calendar' ? "sm" : 0} mt={viewMode === 'calendar' ? "sm" : 0} className={viewMode === 'calendar' ? styles.detailPanel : ''} pos="relative">`);

modifiedDetailPanelBlock = modifiedDetailPanelBlock.trim();

if (modifiedDetailPanelBlock.startsWith('{selectedCompetition && (')) {
    modifiedDetailPanelBlock = modifiedDetailPanelBlock.substring('{selectedCompetition && ('.length).trim();
}
if (modifiedDetailPanelBlock.endsWith(')}')) {
    modifiedDetailPanelBlock = modifiedDetailPanelBlock.substring(0, modifiedDetailPanelBlock.length - 2).trim();
}

const returnStart = content.indexOf('    return (\\n        <Container');
if (returnStart === -1) {
    console.error("Could not find return start");
    process.exit(1);
}

let newContent = content.substring(0, returnStart) + 
    '    const detailPanelContent = selectedCompetition && (\\n        ' + modifiedDetailPanelBlock + '\\n    );\\n\\n' +
    content.substring(returnStart, detailPanelStart) +
    `            {viewMode === 'calendar' && detailPanelContent}\\n` +
    content.substring(blockEnd);

fs.writeFileSync(path, newContent);
console.log("Extraction successful");
